Event tracker repository implementation

9sneha-n opened the pull request:
### :pushpin: References

-   **Issue:** Closes #? https://app.clickup.com/t/8694tp2rg, #8694tp2rg

### :memo: Implementation
1. Implement DiseaseOutbreakEvent Repo
2. Implement partially OrgUnit Repo
3.  Implement partially Options Repo
4. Implement partially Team Member Repo

### :video_camera: Screenshots/Screen capture


### :fire: Notes to the tester

anagperal commented on src/data/repositories/OptionsD2Repository.ts:
```diff
@@ -0,0 +1,25 @@
+        if (!id) return Future.success({ id: "", name: "" });
+        return apiToFuture(
+            this.api.metadata.get({
+                options: { fields: { id: true, name: true }, filter: { id: { eq: id } } },
```
Instead of id will be code, right? Because mainSyndromeId, suspectedDiseaseId and notificationSourceId are codes (maybe we can change the name to Code instead of Id?)

anagperal commented on src/data/repositories/TeamMemberD2Repository.ts:
```diff
@@ -0,0 +1,46 @@
+            this.api.metadata.get({
+                users: {
+                    fields: { id: true, name: true, email: true, phoneNumber: true },
+                    filter: { username: { eq: id } },
```
Maybe for consistency add `username: true` to fields and to TeamMember because `incidentManagerName` is the username instead of the id. Also it's better to change to  `incidentManagerUsername` ?

9sneha-n replied:
Added username: true , however do not understand the second part of the comment. Incident Manager is a type of role in the Team Member. Other Team Members could have other roles, so having the name incidentManagerUsername would not make sense, right?

anagperal replied:
But I was thinking that `incidentManagerName` in `DiseaseOutbreakEventBaseAttrs` was the incident manager username selected in the form so I was populating this property with the username (tha's also why I need username in `TeamMember` entity, and also needed to populate the options in te incident manager selector)

anagperal replied:
By the way, I tried to do it in the other PR but I had an error saying: `Property 'username' does not exist on type 'D2User'`

9sneha-n replied:
ahh ok, i get it now.

anagperal commented on src/data/repositories/utils/DiseaseOutbreakMapper.ts:
```diff
@@ -0,0 +1,240 @@
+                date: new Date(
+                    getValueFromMap("initiatePublicHealthCounterMeasuresNA", trackedEntity)
+                ),
+                na:
```
I think that all na fields when `true` in DHIS2 the app should show `false`

anagperal commented on src/data/repositories/utils/DiseaseOutbreakMapper.ts:
```diff
@@ -0,0 +1,240 @@
+            },
+            initiatePublicHealthCounterMeasures: {
+                date: new Date(
+                    getValueFromMap("initiatePublicHealthCounterMeasuresNA", trackedEntity)
```
"initiatePublicHealthCounterMeasuresDate"

anagperal commented on src/data/repositories/utils/DiseaseOutbreakMapper.ts:
```diff
@@ -0,0 +1,240 @@
+                    getValueFromMap("initiatePublicHealthCounterMeasuresNA", trackedEntity)
+                ),
+                na:
+                    getValueFromMap("initiatePublicHealthCounterMeasuresDate", trackedEntity) ===
```
"initiatePublicHealthCounterMeasuresNA"

anagperal reviewed (changes requested):
thanks @9sneha-n!! The idea of the mappers is very good! I have added some comments in the code

delemaf reviewed (approved):
So for it looks good tome, mapping is a very nice idead 👏

anagperal reviewed (changes requested):
thanks @9sneha-n!! All good!! we'll wait to comment the `username` error in the tech meeting and just one more thing related to how is modeled the `na` fields in DHIS2 and how are in the app: https://github.com/EyeSeeTea/zebra-dev/pull/6/files#r1705539592

tokland commented on src/data/repositories/utils/DiseaseOutbreakMapper.ts:
```diff
@@ -0,0 +1,252 @@
+function getValueFromDiseaseOutbreak(
+    key: (typeof DiseaseOutbreakCodes)[keyof typeof DiseaseOutbreakCodes],
+    diseaseOutbreak: DiseaseOutbreakEventBaseAttrs
+): string {
```
I can't see a drawback of the mapping approach, but it left me thinking what would be the non-mapping approach :) I think it would be something like this:

```ts
const attributeValues = {
  "RTSL_ZEB_TEA_EVENT_id": diseaseOutbreak.eventId.toString(),
  "RTSL_ZEB_TEA_EVENT_NAME": diseaseOutbreak.name,
  // ...
}
const attributes =getAttributesFromAttributeValues(attributeValues)
```

With both styles at hand, pick the one that looks more declarative to you.

tokland reviewed (commented):
I've added some comments, but 99% very minor things, code looks good.

9sneha-n commented on src/CompositionRoot.ts:
```diff
@@ -1,26 +1,59 @@
+            ),
+            getAll: new GetAllDiseaseOutbreaksUseCase(repositories.diseaseOutbreakEventRepository),
+            save: new SaveDiseaseOutbreakUseCase(repositories.diseaseOutbreakEventRepository),
+        },
```
nice idea! :)

9sneha-n commented on src/data/repositories/OptionsD2Repository.ts:
```diff
@@ -0,0 +1,25 @@
+    constructor(private api: D2Api) {}
+
+    get(code: Code): FutureData<NamedRef> {
+        if (!code) return Future.success({ id: "", name: "" });
```
noted for future

9sneha-n commented on src/data/repositories/test/DiseaseOutbreakEventTestRepository.ts:
```diff
@@ -0,0 +1,55 @@
+    }
+    getConfigStrings(): FutureData<ConfigLabel[]> {
+        throw new Error("Method not implemented.");
+    }
```
have implemented get(), getAll() and save() as corresponding d2 functions are implemented. Will implement delete() and getConfigStrings() later once the d2 function is implemented.

9sneha-n commented on src/domain/repositories/OptionsRepository.ts:
```diff
@@ -0,0 +1,6 @@
+import { FutureData } from "../../data/api-futures";
+import { Id, NamedRef } from "../entities/Ref";
+
+export interface OptionsRepository {
```
Will check with Jorge, in the meanwhile, i will give an alias to the key value pair.

xurxodev replied:
Yes I agree with Arnau

9sneha-n commented on src/domain/repositories/DiseaseOutbreakEventRepository.ts:
```diff
@@ -0,0 +1,10 @@
+    getAll(): FutureData<DiseaseOutbreakEventBaseAttrs[]>;
+    save(diseaseOutbreak: DiseaseOutbreakEventBaseAttrs): FutureData<void>;
+    getConfigStrings(): FutureData<ConfigLabel[]>;
+}
```
will ask this one and the next one (which are related) to Jorge.

xurxodev replied:
Yes a repository should represent the gateway of a domain entity to persistence and a repository always should return the domain entity

9sneha-n commented on src/domain/usecases/GetDiseaseOutbreakByIdUseCase.ts:
```diff
@@ -0,0 +1,65 @@
+    ) {}
+
+    public execute(id: Id): FutureData<DiseaseOutbreakEvent> {
+        return this.diseaseOutbreakRepository.get(id).flatMap(diseaseOutbreakEventBase => {
```
will ask Jorge

xurxodev replied:
I don't have context to have a strong opinion here.

From Clean Architecture perspective, It makes sense that a use case need to call to more of a repository if it's necessary. For example, updateOrderUseCase need call to userRepository to know if user have permissions to update orders and then call to orderRepository.

But in this case, I don't know if the repositories are correct or not.

9sneha-n commented:
> thanks @9sneha-n!! All good!! we'll wait to comment the `username` error in the tech meeting and just one more thing related to how is modeled the `na` fields in DHIS2 and how are in the app: https://github.com/EyeSeeTea/zebra-dev/pull/6/files#r1705539592

Thanks @anagperal Good point, its fixed now

9sneha-n commented:
> [code-only review] I've added some comments, but 99% very minor things, code looks good.

Thanks @tokland , addressed all the comments, except the ones requiring Jorge's input. Will ask him over slack or in technical decision meeting

tokland commented on src/data/repositories/utils/AssertOrError.ts:
```diff
@@ -0,0 +1,7 @@
+import { Future } from "../../../domain/entities/generic/Future";
+import { FutureData } from "../../api-futures";
+
+export function assertOrError<T>(objects: T, name: string): FutureData<NonNullable<T>> {
```
singular: objects -> obj

tokland commented on src/webapp/pages/dashboard/DashboardPage.tsx:
```diff
@@ -38,11 +38,7 @@ export const DashboardPage: React.FC = React.memo(() => {
-        compositionRoot.diseaseOutbreakEvent.getAll,
-        compositionRoot.diseaseOutbreakEvent.save,
-    ]);
+    }, [compositionRoot.diseaseOutbreakEvent]);
```
Yes. Even `compositionRoot`, as the whole object is immutable.

tokland commented on src/data/repositories/DiseaseOutbreakEventD2Repository.ts:
```diff
@@ -22,11 +23,16 @@ export class DiseaseOutbreakEventD2Repository implements DiseaseOutbreakEventRep
+        )
+            .flatMap(
+                (response): FutureData<D2TrackerTrackedEntity> =>
+                    assertOrError(response.instances[0], "Tracked entity")
```
In this case the type of the block is redundant, I'd remove it

tokland commented on src/data/repositories/consts/DiseaseOutbreakConstants.ts:
```diff
@@ -33,3 +39,102 @@ export const DiseaseOutbreakCodes = {
+        "Environmental",
+        "Unknown",
+    ].includes(hazardType);
+}
```
This is error-prone. You may write some string wrong and TS will follow along. The orthodox+single-truth way of doing is, in the entity:

```ts
const hazardTypes = [
    "Biological:Human",
    "Biological:Animal",
    "Chemical",
    "Environmental",
    "Unknown",
] as const;

export type HazardType = typeof hazardTypes[number];
```

You see, the other way around, you start from values and extract types from it. And now you can `hazardTypes` (useful for type validations, dropdown, and so on)

tokland commented on src/data/repositories/consts/DiseaseOutbreakConstants.ts:
```diff
@@ -1,3 +1,9 @@
+
 export const RTSL_ZEBRA_PROGRAM_ID = "qkOTdxkte8V";
 export const RTSL_ZEBRA_ORG_UNIT_ID = "PS5JpkoHHio";
 export const RTSL_ZEBRA_TRACKED_ENTITY_TYPE_ID = "lIzNjLOUAKA";
```
(some lines below) DiseaseOutbreakCodes -> diseaseOutbreakCodes

tokland commented on src/data/repositories/consts/DiseaseOutbreakConstants.ts:
```diff
@@ -33,3 +39,102 @@ export const DiseaseOutbreakCodes = {
+export type KeyCode = (typeof DiseaseOutbreakCodes)[keyof typeof DiseaseOutbreakCodes];
+
+export function isStringInDiseaseOutbreakCodes(code: string): code is KeyCode {
+    return Object.values(DiseaseOutbreakCodes).includes(code as KeyCode);
```
[nit-picking] the orthodox way is casting the other way (which is safe): `(Object.values(DiseaseOutbreakCodes) as string[]).includes(code)`

tokland commented on src/data/repositories/consts/DiseaseOutbreakConstants.ts:
```diff
@@ -33,3 +39,102 @@ export const DiseaseOutbreakCodes = {
+        RTSL_ZEB_TEA_SPECIFY_DATE2:
+            diseaseOutbreak.earlyResponseActions.appropriateCaseManagement.date?.toISOString() ??
+            "",
+        RTSL_ZEB_TEA_APPROPRIATE_PUBLIC_HEALTH: diseaseOutbreak.earlyResponseActions
```
I'd still abstract these two, the toISOString, and the "true", and they are used many times.

tokland commented on src/data/repositories/consts/DiseaseOutbreakConstants.ts:
```diff
@@ -33,3 +39,102 @@ export const DiseaseOutbreakCodes = {
+        case "Unknown":
+            return "UNKNOWN";
+    }
+}
```
Not that it's wrong, but note that a simple object `Record<HazardType, string>`.

tokland reviewed (commented):
Nice! Very close. Sorry for the nit-picking! :)

anagperal reviewed (approved):
Looks good to me!! thanks @9sneha-n !! :rocket:

anagperal commented on src/data/repositories/utils/DiseaseOutbreakMapper.ts:
```diff
@@ -0,0 +1,181 @@
+        created: trackedEntity.createdAt ? new Date(trackedEntity.createdAt) : new Date(),
+        lastUpdated: trackedEntity.updatedAt ? new Date(trackedEntity.updatedAt) : new Date(),
+        createdByName: undefined,
+        hazardType: getHazardTypeValue(getValueFromMap("hazardType", trackedEntity)),
```
This does not seem to work because the hazard types in the trackedEntity are like `BIOLOGICAL_ANIMAL`, `BIOLOGICAL_HUMAN` ... (check https://metadata.eyeseetea.com/rsl-zebra/dhis-web-maintenance/index.html#/edit/otherSection/optionSet/p2N1GkG713S/options) so first we need to map from these codes to  "Biological:Animal",    "Biological:Human",  "Chemical",   "Environmental" or "Unknown". Also there is one new: Biological: Human and Animal

anagperal commented on src/data/repositories/consts/DiseaseOutbreakConstants.ts:
```diff
@@ -0,0 +1,140 @@
+            diseaseOutbreak.earlyResponseActions.initiateInvestigation.toISOString(),
+        RTSL_ZEB_TEA_CONDUCT_EPIDEMIOLOGICAL_ANALYSIS:
+            diseaseOutbreak.earlyResponseActions.conductEpidemiologicalAnalysis.toISOString(),
+        RTSL_ZEB_TEA_LABORATORY_CONFIRMATION: diseaseOutbreak.earlyResponseActions
```
For NA values is the other way arround:

```
RTSL_ZEB_TEA_LABORATORY_CONFIRMATION: diseaseOutbreak.earlyResponseActions
            .laboratoryConfirmation.na
            ? ""
            : "true"
```

9sneha-n replied:
Hi @anagperal In DHIS all of these are mandatory tracked entity attribute of Yes only type. They cannot be null or empty. The app implementation needs to take care of the true value, we can discuss further over call.

anagperal reviewed (changes requested):
thansk @9sneha-n ! I had approved the PR but in using it I noticed a couple of small bugs :pray:

tokland reviewed (approved):
All good as far as I can see, code-wise!

anagperal reviewed (approved):
Thanks @9sneha-n !!
